app/models/carto/map.rb

```diff
@@ -158,7 +158,12 @@ def admits_layer?(layer)
   end
 
   def can_add_layer?(user)
-    return false if user.max_layers && user.max_layers <= data_layers.count
+    emptyDatasetName = Cartodb.config[:shared_empty_dataset_name]
+    layers_on_map = data_layers
+    layer_count = layers_on_map.count
+    layer_count -= 1 if layers_on_map.any?{ |layer| layer.options[:table_name] == emptyDatasetName }
+
+    return false if user.max_layers && user.max_layers <= layer_count
```

Comment:
Hey I think you can rewrite this as:

```ruby
empty_dataset_name = Cartodb.config[:shared_empty_dataset_name]

countable_layers = data_layers.reject do |layer|
  layer.options[:table_name] == empty_dataset_name
end

return false if user.max_layers && user.max_layers <= countable_layers.count
```

See if you like it more (It's a bit more compact).

Reply:
existing one is more optimal, so I will keep it.

Reply:
Just curious: "optimal" how? Readability or performance?

If you're worried about performance, I don't think it will make much of a difference :)

Reply:
performance. `any?` should return once first one found (empty dataset is normally the first dataset anyway). thats why should perform better. reject should be iterating even though an instance found. not a major performance gain since map will not have 10s of datasets.

Reply:
Fair enough.. in these situations I'd usually go for readability.

Your call, up to you! It's good either way